lib/virtual-net/src/ruleset.rs

```diff
@@ -0,0 +1,1388 @@
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr};
+use std::ops::RangeInclusive;
+use std::str::FromStr;
+use std::sync::{Arc, RwLock};
+
+use ipnet::{Ipv4Net, Ipv6Net};
+use iprange::IpRange;
+
+// <rule_kind>:<rule_action>=<rule_expr>
+//
+// <rule_kind>: dns, ipv4, ipv6
+//
+// <rule_action>: allow | deny
+//
+// dns:
+// <rule_expr>:
+// {<domain_spec>}:{<port_spec>} (this will be expanded to an outbound IP rule)
+// <domain_spec>: domain | domain glob | *
+//
+// ipv4:
+// <rule_expr>:
+// {<ipv4_spec>}:{<port_spec>}/<in|out>
+// <ipv4_spec>: ipv4 | ipv4_range | *
+//
+// ipv6:
+// <rule_expr>:
+// {<ipv6_spec>}:{<port_spec>}/<in|out>
+// <ipv4_spec>: ipv6 | ipv6_range | *
+//
+// <port_spec>: port | start_port-end_port | *
+
+#[derive(Debug, thiserror::Error)]
+pub enum RulesetError {
+    #[error("invalid connection direction: {0}")]
+    DirectionParsingError(String),
```

Comment:
```suggestion
    Direction(String),
```

No need to repeat Error in the variants, also, all of them are parse errors.